archive/whilecc.lean

```diff
@@ -0,0 +1,1250 @@
+import algebra
+import data.real.basic
+import data.vector
+import tactic.explode
+import tactic.find
+import tactic.induction
+import tactic.linarith
+import tactic.rcases
+import tactic.rewrite
+import tactic.ring_exp
+import tactic.tidy
+import tactic.where
+
+set_option pp.beta true
+set_option pp.coercions true
+set_option pp.generalized_field_notation false
+set_option trace.check true
+set_option pp.notation true
+/-!
+This is a compiler and proof that is translated from the second part of Concrete Semantics. http://concrete-semantics.org/
+
+The proof consists of just the forward part of the proof presented in Concrete
+Semantics, i.e. it's a partial correctness proof. It proves that the compiler is
+correct for all terminating programs.
+
+-/
+/- # LoVe Library
+ Copied from https://github.com/blanchette/logical_verification_2020/
+ -/
+namespace LoVe
+namespace rtc
+
+inductive star {α : Sort*} (r : α → α → Prop) (a : α) : α → Prop
+| refl {}    : star a
+| tail {b c} : star b → r b c → star c
+
+attribute [refl] star.refl
+
+namespace star
+
+variables {α : Sort*} {r : α → α → Prop} {a b c d : α}
+
+@[trans] lemma trans (hab : star r a b) (hbc : star r b c) :
+  star r a c :=
+begin
+  induction' hbc,
+  case refl {
+    assumption },
+  case tail : c d hbc hcd hac {
+    exact (tail (hac hab)) hcd }
+end
+
+lemma single (hab : r a b) :
+  star r a b :=
+refl.tail hab
+end star
+end rtc
+
+def state :=
+string → int
+
+def state.update (name : string) (val : int) (s : state) : state :=
+λname', if name' = name then val else s name'
+
+notation s `{` name ` ↦ ` val `}` := state.update name val s
+
+instance : has_emptyc state :=
+{ emptyc := λ_, 0 }
+
+export rtc
+end LoVe
+
+open LoVe
+
+namespace bool
+lemma or_intro (b : bool) : b = tt ∨ b = ff :=
+begin
+  cases' em (b = ff),
+  {
+    apply or.intro_right, exact h,
+  },
+  { simp at h,
+    apply or.intro_left,
+    exact h,
+  },
+end
+end bool
+
+lemma le_iff_not_lt {a b : int} :  a ≤ b -> ¬ b < a :=
```

Comment:
This should be called `not_lt_of_le`, as it's not an `iff` at all!

Having said that, mathlib already has this lemma, `library_search` should find it.